# InfoHiC
- InfoHiC is a method to predict 3D genome folding from complex DNA rearrangements, which enables the training and validation usage of cancer Hi-C data. InfoHiC provides a decomposed level of interaction views of multiple contigs from the cancer Hi-C matrix. InfoHiC is composed of convolutional neural network (CNNs) that outputs chromatin interactions of genomic contigs and merges them in the total Hi-C views, based on a haplotype-specific copy number (HSCN) encoding that represent genomic variants in the contig matrix.

<p align="center">
    <img width="1500" src="https://github.com/DMCB-GIST/InfoHiC/blob/main/doc/overview.png">
  </a>
</p>

# Requirements
- R (version 3.4.3)
    - plyr
    - dplyr
    - RColorBrewer
    - reshape2
    - ggplot2
    - ggforce
    - grid
    - SepctralTAD
- python (version 3)
    - tensorflow (version 1.14.0)
    - numpy
    - pysam
    - pickle
    - scipy
    - neoloop
    - iced
- perl
    - Bio::DB::Fasta;

# Environment settings
```
export InfoHiC_lib=/home/dmcblab/InfoHiC
export PATH=$InfoHiC_lib/InfoGenomeR_processing:$InfoHiC_lib/InfoHiC_tensorflow:$InfoHiC_lib/post_analysis:$PATH
```

# Inputs
- Following inputs are required, which can be generated by InfoGenomeR (https://github.com/dmcblab/InfoGenomeR)
	- contigs.fa
	- contigs.index
	- contigs.bed
	- contigs.cn_file
	- sv_window
	- ref_window
# Outputs
- contig Hi-C matrix and total Hi-C matrix
- neoloop and neo-TAD annotation
- Enhancer hijacking annotation


# Running InfoHiC
- InfoGenomeR_processing
```
Usage: InfoGenomeR_processing <InfoGenomeR_output> <fasta_file>

Options:
        -e, --ens (required)
                 Ensembl annotation
        -s, --se (required)
                 Super enhancer annotation
        -t, --te (required)
                 Typical enhancer annotation
        -h, --help
```

- InfoHiC_training
```
Usage: InfoHiC_training <InfoGenomeR_output> <cancer_HiC_matrix> [options]

Options:
```

- InfoHiC_test
```
Usage: InfoHiC_test <InfoGenomeR_output> [options]

Options:
        -m, --mode (required)
                 Select the mode (HSCN_encoding, HSCN_decoding)
        -w, --window (required)
                 window size (1Mb, 2Mb)
        -g, --gpu (required)
                 gpu index (default: 0)
        -c, --checkpoint (required)
                 trained model checkpoint

        -h, --help
```
- InfoHiC post analysis
```
Usage: InfoHiC_post <InfoHiC_output> [options]

Options:
        -w, --window (required)
                 window size (1Mb, 2Mb)
        -m, --true_matrix (optional)
                 comparison with the true matrix. Enter a true matrix directory
        -h, --help
```

# Tutorial 1 (T47D)
```
wget https://zenodo.org/T47D_InfoGenomeR.tar.gz
tar -xvf T47D_InfoGenomeR.tar.gz
cd T47D_InfoGenomeR
```
- Process the InfoGenomeR output
```
### Process the InfoGenomeR output 
InfoGenomeR_processing output -e ${InfoHiC_lib}/humandb/hg19_ensGene.txt -s SE_package.bed.BRCA.rf -t TE_package.bed.BRCA.rf
```
- Predict Hi-C matrix
```
### Predict Hi-C matrix
InfoHiC_test InfoGenomeR_output -m HSCN_encoding -w 2Mb -g 0 -c ${InfoHiC_lib}/models/T47D_transfer_2Mb/best_checkpoint
```

- perform a post analysis
```
### post analysis
wget https://zenodo.org/T47D_experiment.tar.gz
tar -xvf T47D_experiment.tar.gz
InfoHiC_post InfoGenomeR_output -w 2Mb -m T47D_experiment
```

<p align="center">
    <img width="400" src="https://github.com/dmcb-gist/InfoHiC/blob/main/doc/euler.8.14/karyotypes.png">
  </a>
</p>

<p align="center">
    <img width="700" src="https://github.com/dmcb-gist/InfoHiC/blob/main/doc/euler.8.14/HiC_prediction.png">
  </a>
</p>
<p align="center">
    <img width="700" src="https://github.com/dmcb-gist/InfoHiC/blob/main/doc/euler.8.14/contig_HiC.png">
  </a>
</p>

